Improve Canvas ImageData noise injection algorithm

sysrqb opened the pull request:
#### d89e7f323029abc0d06d022242485dcf44b39979
<pre>
Improve Canvas ImageData noise injection algorithm
<a href="https://bugs.webkit.org/show_bug.cgi?id=257574">https://bugs.webkit.org/show_bug.cgi?id=257574</a>
rdar://109271231

Reviewed by Kimmo Kinnunen.

The current algorithm attempts to recognize gradients within a PixelBuffer, but
the logic for detecting a gradient is incorrect where it assumes a gradient
region always contain strictly increase (or decreasing) colors. This assumption
resulted in poor gradient detection and very visible anomalies in some cases.
In addition, in the case where the gradient was correctly detected, in some
cases the amount of added noise was too large and it cause a clustering effect
around some colors within the resulting image.

In this patch, I modified the noise-addition algorithm such that it provides
the best result for every pixel depending on its surrounding colors, instead of
restricting special behavior to gradients. We introduce a concept of &quot;related
colors&quot; which identifies two colors as related if the difference between each
of their RGBA components is not more than 8 - this threshold was mostly chosen
by experimentation.

The algorithm proceeds as follows:
1) For each color, identify if any of the immediately surrounding colors are related.
2) If any of the colors are related, then find the colors with the smallest distance.
3) If any related colors were identified, then those colors define the upper/lower
   bounds within which noise can be added.
4) Furthermore, if any related colors were identified, then noise is restricted
   to +/-1 (~0.5%), otherwise it is restricted to +/-3 (~1%);
5) After identifying the closest surrounding colors, we define the upper and lower
   bounds for each color component.
6) Finally, each color component is mutated by adding the noise value

* Source/WebCore/html/CanvasNoiseInjection.cpp:
(WebCore::setTightnessBounds):
(WebCore::boundingNeighbors):
(WebCore::CanvasNoiseInjection::postProcessDirtyCanvasBuffer):
(WebCore::lowerAndUpperBound):
(WebCore::adjustNeighborColorBounds):
(WebCore::CanvasNoiseInjection::postProcessPixelBufferResults const):
(WebCore::getGradientNeighbors): Deleted.
* Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/canvas-fingerprinting.html:

Canonical link: <a href="https://commits.webkit.org/265333@main">https://commits.webkit.org/265333@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb3d14fe2fd7b12bcb8779f37f0e8acc0e6e4eb6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10470 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10699 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10974 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12110 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10042 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10482 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12969 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10655 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12989 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10629 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11508 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8807 "53 api tests failed or timed out") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12509 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8805 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9463 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16723 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9884 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9615 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12863 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10065 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8155 "1 flakes 4 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9224 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2543 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13478 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9926 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer reviewed (approved):
rs=me based on Kimmo's previous review.

sysrqb committed:
Improve Canvas ImageData noise injection algorithm
https://bugs.webkit.org/show_bug.cgi?id=257574
rdar://109271231

Reviewed by Kimmo Kinnunen.

The current algorithm attempts to recognize gradients within a PixelBuffer, but
the logic for detecting a gradient is incorrect where it assumes a gradient
region always contain strictly increase (or decreasing) colors. This assumption
resulted in poor gradient detection and very visible anomalies in some cases.
In addition, in the case where the gradient was correctly detected, in some
cases the amount of added noise was too large and it cause a clustering effect
around some colors within the resulting image.

In this patch, I modified the noise-addition algorithm such that it provides
the best result for every pixel depending on its surrounding colors, instead of
restricting special behavior to gradients. We introduce a concept of "related
colors" which identifies two colors as related if the difference between each
of their RGBA components is not more than 8 - this threshold was mostly chosen
by experimentation.

The algorithm proceeds as follows:
1) For each color, identify if any of the immediately surrounding colors are related.
2) If any of the colors are related, then find the colors with the smallest distance.
3) If any related colors were identified, then those colors define the upper/lower
   bounds within which noise can be added.
4) Furthermore, if any related colors were identified, then noise is restricted
   to +/-1 (~0.5%), otherwise it is restricted to +/-3 (~1%);
5) After identifying the closest surrounding colors, we define the upper and lower
   bounds for each color component.
6) Finally, each color component is mutated by adding the noise value

* Source/WebCore/html/CanvasNoiseInjection.cpp:
(WebCore::setTightnessBounds):
(WebCore::boundingNeighbors):
(WebCore::CanvasNoiseInjection::postProcessDirtyCanvasBuffer):
(WebCore::lowerAndUpperBound):
(WebCore::adjustNeighborColorBounds):
(WebCore::CanvasNoiseInjection::postProcessPixelBufferResults const):
(WebCore::getGradientNeighbors): Deleted.
* Tools/TestWebKitAPI/Tests/WebKit/AdvancedPrivacyProtections.mm:
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/canvas-fingerprinting.html:

Canonical link: https://commits.webkit.org/265333@main
